File: physicar_bringup/scripts/scan_filter_node.py
#!/usr/bin/env python3
"""
Scan Filter Node for PhysiCar

Filters invalid LaserScan readings (inf, nan, out of range) to improve
downstream odometry estimation (rf2o_laser_odometry).

Subscribes: /scan
Publishes: /scan_filtered
"""

import math

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy
from sensor_msgs.msg import LaserScan


class ScanFilterNode(Node):
    def __init__(self):
        super().__init__('scan_filter')
        
        # Parameters
        self.declare_parameter('input_topic', '/scan')
        self.declare_parameter('output_topic', '/scan_filtered')
        self.declare_parameter('range_margin', 0.1)  # margin from range_max
        
        input_topic = self.get_parameter('input_topic').value
        output_topic = self.get_parameter('output_topic').value
        self.range_margin = self.get_parameter('range_margin').value
        
        # QoS for lidar (best effort, volatile)
        sensor_qos = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.VOLATILE,
            depth=1
        )
        
        # Subscriber and Publisher
        self.subscription = self.create_subscription(
            LaserScan,
            input_topic,
            self.scan_callback,
            sensor_qos
        )
        
        self.publisher = self.create_publisher(
            LaserScan,
            output_topic,
            sensor_qos
        )
        
        self.get_logger().info(f'Scan filter: {input_topic} -> {output_topic}')
    
    def scan_callback(self, msg: LaserScan):
        """Filter invalid readings and republish."""
        # Create filtered message (copy metadata)
        filtered = LaserScan()
        filtered.header = msg.header
        filtered.angle_min = msg.angle_min
        filtered.angle_max = msg.angle_max
        filtered.angle_increment = msg.angle_increment
        filtered.time_increment = msg.time_increment
        filtered.scan_time = msg.scan_time
        filtered.range_min = msg.range_min
        filtered.range_max = msg.range_max
        
        # Filter ranges
        # rf2o treats 0 as invalid, so convert bad readings to 0
        effective_max = msg.range_max - self.range_margin
        filtered_ranges = []
        
        for r in msg.ranges:
            if (math.isfinite(r) and 
                r > msg.range_min and 
                r < effective_max):
                filtered_ranges.append(r)
            else:
                filtered_ranges.append(0.0)
        
        filtered.ranges = filtered_ranges
        
        # Copy intensities if present
        if msg.intensities:
            filtered.intensities = list(msg.intensities)
        
        self.publisher.publish(filtered)


def main(args=None):
    rclpy.init(args=args)
    node = ScanFilterNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: physicar_bringup/src/scan_filter_node.cpp
/**
 * Scan Filter Node for PhysiCar (C++ port)
 *
 * Filters invalid LaserScan readings (inf, nan, out of range) to improve
 * downstream odometry estimation (rf2o_laser_odometry).
 *
 * Subscribes: /scan
 * Publishes:  /scan_filtered
 */

#include <cmath>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>

class ScanFilterNode : public rclcpp::Node
{
public:
  ScanFilterNode() : Node("scan_filter")
  {
    declare_parameter("input_topic", std::string("/scan"));
    declare_parameter("output_topic", std::string("/scan_filtered"));
    declare_parameter("range_margin", 0.1);

    auto input_topic = get_parameter("input_topic").as_string();
    auto output_topic = get_parameter("output_topic").as_string();
    range_margin_ = get_parameter("range_margin").as_double();

    // Sensor QoS (best effort, volatile, depth 1)
    auto sensor_qos = rclcpp::SensorDataQoS().keep_last(1);

    pub_ = create_publisher<sensor_msgs::msg::LaserScan>(output_topic, sensor_qos);
    sub_ = create_subscription<sensor_msgs::msg::LaserScan>(
      input_topic, sensor_qos,
      [this](const sensor_msgs::msg::LaserScan::SharedPtr msg) { scan_callback(msg); });

    RCLCPP_INFO(get_logger(), "Scan filter: %s -> %s", input_topic.c_str(), output_topic.c_str());
  }

private:
  void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
  {
    auto filtered = std::make_unique<sensor_msgs::msg::LaserScan>();
    filtered->header = msg->header;
    filtered->angle_min = msg->angle_min;
    filtered->angle_max = msg->angle_max;
    filtered->angle_increment = msg->angle_increment;
    filtered->time_increment = msg->time_increment;
    filtered->scan_time = msg->scan_time;
    filtered->range_min = msg->range_min;
    filtered->range_max = msg->range_max;

    const float effective_max = msg->range_max - static_cast<float>(range_margin_);
    const size_t n = msg->ranges.size();
    filtered->ranges.resize(n);

    for (size_t i = 0; i < n; ++i) {
      const float r = msg->ranges[i];
      filtered->ranges[i] =
        (std::isfinite(r) && r > msg->range_min && r < effective_max) ? r : 0.0f;
    }

    if (!msg->intensities.empty()) {
      filtered->intensities = msg->intensities;
    }

    pub_->publish(std::move(filtered));
  }

  rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr pub_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_;
  double range_margin_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ScanFilterNode>());
  rclcpp::shutdown();
  return 0;
}
